### PR TITLE
planner: support converting `json_overlaps/contains` to IndexMerge to access MVIndex

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #40191

Problem Summary: planner: support converting `json_overlaps/contains` to IndexMerge to access MVIndex

### What is changed and how it works?

wait for https://github.com/pingcap/tidb/pull/40175

planner: support converting `json_overlaps/contains` to IndexMerge to access MVIndex

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/indexmerge_path.go [hunk 1]
```diff
@@ -560,10 +563,16 @@ func (ds *DataSource) generateIndexMergeJSONMVIndexPath(normalPathCnt int, filte
 					continue
 				}
 				vals = append(vals, v)
-			case ast.JSONOverlaps: // (json_overlaps(a->'$.zip', '[1, 2, 3]')
-				continue // TODO: support json_overlaps
 			case ast.JSONContains: // (json_contains(a->'$.zip', '[1, 2, 3]')
-				continue // TODO: support json_contains
+				indexMergeIsIntersection = true
+				fallthrough
+			case ast.JSONOverlaps: // (json_overlaps(a->'$.zip', '[1, 2, 3]')
```
Comment (xiongjiwei) [inline planner/core/indexmerge_path.go:569]: `json_overlaps(a, b)` is equal to `json_overlaps(b, a)`, that should be considered

Comment (qw4990) [reply planner/core/indexmerge_path.go]: Updated. PTAL @xiongjiwei

**Review by xiongjiwei — APPROVED** [review]

File: planner/core/testdata/index_merge_suite_out.json [hunk 2]
```diff
@@ -49,6 +49,128 @@
...
+          "├─IndexRangeScan(Build) 10.00 cop[tikv] table:t, index:j0_0(cast(json_extract(`j0`, _utf8mb4'$.path0') as signed array)) range:[1,1], keep order:false, stats:pseudo",
+          "├─IndexRangeScan(Build) 10.00 cop[tikv] table:t, index:j0_0(cast(json_extract(`j0`, _utf8mb4'$.path0') as signed array)) range:[2,2], keep order:false, stats:pseudo",
+          "├─IndexRangeScan(Build) 10.00 cop[tikv] table:t, index:j0_0(cast(json_extract(`j0`, _utf8mb4'$.path0') as signed array)) range:[3,3], keep order:false, stats:pseudo",
+          "└─TableRowIDScan(Probe) 0.00 cop[tikv] table:t keep order:false, stats:pseudo"
+        ]
+      },
+      {
+        "SQL": "select /*+ use_index_merge(t, j0_0) */ * from t where json_overlaps((j0->'$.path0'), '[1, 2, 3]')",
+        "Plan": [
+          "Selection 0.00 root  json_overlaps(json_extract(test.t.j0, \"$.path0\"), cast(\"[1, 2, 3]\", json BINARY))",
```
Comment (time-and-fate) [inline planner/core/testdata/index_merge_suite_out.json:66]: It looks like this `Selection` is redundant.

File: planner/core/indexmerge_path.go [hunk 3]
```diff
@@ -612,12 +634,62 @@ func (ds *DataSource) generateIndexMergeJSONMVIndexPath(normalPathCnt int, filte
...
+		return nil, false
+	}
+	if jsonCast.FuncName.L != ast.Cast && jsonCast.GetType().EvalType() != types.ETJson {
+		return nil, false
+	}
+	if _, isConst := jsonCast.GetArgs()[0].(*expression.Constant); !isConst {
+		return nil, false
+	}
+
+	jsonArray, isNull, err := jsonArrayExpr.EvalJSON(sctx, chunk.Row{})
```
[the remark this answers is not in the document]
Comment (qw4990) [reply planner/core/indexmerge_path.go:658]: If we don't use `EvalJSON` here, we need to convert the array string like `"[1, 2, 3]"` to the corresponding slice like `[1, 2, 3]` manually, which seems dangerous. So just using `EvalJSON + JSON.ArrayGetElem` seems safer and easier.

File: planner/core/testdata/index_merge_suite_out.json (hunk 2 above)
Comment (qw4990) [reply planner/core/testdata/index_merge_suite_out.json:66]: The reason is `ast.JSONOverlaps` cannot be pushed down to TiKV (https://github.com/pingcap/tidb/blob/master/expression/expression.go#L1036), and this let the optimizer keeps an Selection upon the DataSrouce (https://github.com/pingcap/tidb/blob/master/planner/core/rule_predicate_push_down.go#L141), which is expected.

File: planner/core/indexmerge_path.go [hunk 4]
```diff
@@ -612,12 +634,62 @@ func (ds *DataSource) generateIndexMergeJSONMVIndexPath(normalPathCnt int, filte
 				partialPaths = append(partialPaths, partialPath)
 			}
 			indexMergePath := ds.buildIndexMergeOrPath(filters, partialPaths, filterIdx)
+			indexMergePath.IndexMergeIsIntersection = indexMergeIsIntersection
```
[the remark this answers is not in the document]
Comment (qw4990) [reply planner/core/indexmerge_path.go:637]: Now `indexMergeIsIntersection` only works for `ast.JSONContains`. Is it any problem here?

File: planner/core/indexmerge_path.go [hunk 5]
```diff
@@ -612,12 +634,62 @@ func (ds *DataSource) generateIndexMergeJSONMVIndexPath(normalPathCnt int, filte
...
+// jsonArrayExpr2Exprs converts a JsonArray expression to expression list: cast('[1, 2, 3]' as JSON) --> []expr{1, 2, 3}
+func jsonArrayExpr2Exprs(sctx sessionctx.Context, jsonArrayExpr expression.Expression) ([]expression.Expression, bool) {
+	// only support cast(const as JSON)
+	jsonCast, ok := jsonArrayExpr.(*expression.ScalarFunction)
+	if !ok {
+		return nil, false
+	}
+	if jsonCast.FuncName.L != ast.Cast && jsonCast.GetType().EvalType() != types.ETJson {
+		return nil, false
+	}
```
[the remark this answers is not in the document]
Comment (qw4990) [reply planner/core/indexmerge_path.go]: You are right. I've updated this place.

File: planner/core/indexmerge_path.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (time-and-fate) [reply planner/core/indexmerge_path.go:637]: From its name, `buildIndexMergeOrPath` is for the `OR`, not `AND`.

File: planner/core/indexmerge_path.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (time-and-fate) [reply planner/core/indexmerge_path.go:655]: Got it. What about the second question?

File: planner/core/testdata/index_merge_suite_out.json (hunk 2 above)
Comment (time-and-fate) [reply planner/core/testdata/index_merge_suite_out.json:66]: But the `json_overlaps` has already been handled by the `IndexMerge`.
And `json_contains` also doesn't have the `Selection`.

Comment (qw4990) [reply planner/core/testdata/index_merge_suite_out.json:66]: YES, but it's another problem. I'll submit another PR to fix it after.

File: planner/core/indexmerge_path.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (qw4990) [reply planner/core/indexmerge_path.go:658]: OK, I use `IsMutableEffectsExpr` to check it here.

**Review by Yisaer — APPROVED** [review]

**Pull request merged**